source/borderrouter_ws.c

```diff
@@ -0,0 +1,606 @@
+  *  - SYSTEM_TIMER event: For RF interface Handshake purpose
+  *
+  */
+static void borderrouter_tasklet(arm_event_s *event)
+{
+    arm_library_event_type_e event_type;
+    event_type = (arm_library_event_type_e)event->event_type;
+
+    switch (event_type) {
+        case ARM_LIB_NWK_INTERFACE_EVENT:
+
+            if (event->event_id == ws_br_handler.eth_interface_id) {
+                network_interface_event_handler(event);
+            } else {
+                wisun_interface_event_handler(event);
+            }
+
+            break;
+        // comes from the backhaul_driver_init.
+        case APPLICATION_EVENT:
+            if (event->event_id == NR_BACKHAUL_INTERFACE_PHY_DRIVER_READY) {
+                int8_t net_backhaul_id = (int8_t) event->event_data;
+
+                if (backhaul_interface_up(net_backhaul_id) != 0) {
+                    tr_debug("Backhaul bootstrap start failed");
+                } else {
+                    tr_debug("Backhaul bootstrap started");
+                }
+            } else if (event->event_id == NR_BACKHAUL_INTERFACE_PHY_DOWN) {
+                if (backhaul_interface_down() != 0) {
+                } else {
+                    tr_debug("Backhaul interface is down");
+                }
+            }
+            break;
+
+        case ARM_LIB_TASKLET_INIT_EVENT:
+            br_tasklet_id = event->receiver;
+            eth_network_data_init();
+            backhaul_driver_init(borderrouter_backhaul_phy_status_cb);
+            mesh_network_up();
+            eventOS_event_timer_request(9, ARM_LIB_SYSTEM_TIMER_EVENT, br_tasklet_id, 20000);
+            break;
+
+        case ARM_LIB_SYSTEM_TIMER_EVENT:
+            eventOS_event_timer_cancel(event->event_id, event->receiver);
+
+            if (event->event_id == 9) {
+#if MBED_CONF_APP_DEBUG_TRACE
+                arm_print_routing_table();
+                arm_print_neigh_cache();
+                print_memory_stats();
+                // Trace interface addresses. This trace can be removed if nanostack prints added/removed
+                // addresses.
+                print_interface_addresses();
+#endif
+                eventOS_event_timer_request(9, ARM_LIB_SYSTEM_TIMER_EVENT, br_tasklet_id, 20000);
+            }
+            break;
+
+        default:
+            break;
+    }
+}
+
+static void borderrouter_backhaul_phy_status_cb(uint8_t link_up, int8_t driver_id)
+{
+    arm_event_s event = {
+        .sender = br_tasklet_id,
+        .receiver = br_tasklet_id,
+        .priority = ARM_LIB_MED_PRIORITY_EVENT,
+        .event_type = APPLICATION_EVENT,
+        .event_data = driver_id
+    };
+
+    if (link_up) {
+        event.event_id = NR_BACKHAUL_INTERFACE_PHY_DRIVER_READY;
+    } else {
+        event.event_id = NR_BACKHAUL_INTERFACE_PHY_DOWN;
+    }
+
+    eventOS_event_send(&event);
+}
+
+// ethernet interface
+static void network_interface_event_handler(arm_event_s *event)
+{
```

Comment:
Do we have copy paste code for all ethernet handling in all three border routers?
only one handler should be enough for all mesh interfaces
Also one tasklet with hooks calling different mesh interface handlers sounds like a good idea

Reply:
No, only for thread and wi-sun. Agree that these should be cleaned and refactored.